CONTRIBUTING.md

```diff
@@ -207,6 +207,96 @@ When updating dependencies, please only update the particular dependencies you
+message will occur frequently (e.g. a few times per minute), try to find ways to
+either eliminate that message or reduce its frequency (e.g. only logging every
+Nth message, or a summary message every minute or hour).
+
+### Log levels
+
+Different log levels should target different groups of users. At present, only
+**Debug**, **Info** and **Error** levels are supported.
+
+- **Debug**: Should primarily target consensus engine developers (i.e. core team
+  members and developers working on CometBFT forks).
+- **Info** and **Error**: Should primarily target operators and application
+  developers.
+
+### Log messages
+
+Log messages should always be tailored to the intended target audience. Unlike
+Go errors, **log messages must have the first letter of the message
+capitalized**. Only _errors_ in Go should start with a lowercase letter because
+they may end up being chained/embedded, but log messages are not chained in this
+same way.
+
+### Logging parameters
+
+In general, log messages should contain the **bare minimum** amount of
+information for those messages to be actionable by the target audience. So
+instead of dumping large quantities of raw data into the logs (e.g. RPC
+responses, transactions, block data), include lightweight references to the data
+that users can go look up via the RPC, CLI tools, etc. if they are genuinely
+interested in the details.
+
+When outputting variables, also keep in mind **concurrency concerns** of doing
+so. If outputting a pointer, understand that the value associated with that
+pointer at the time of calling the log function may differ to its value at the
+time it is finally serialized into the log message. Perhaps consider creating a
+temporary copy of the specific value you want to output and logging that
+temporary value. Also keep in mind potential data races when doing so.
+
+Finally, **use expensive operations like `fmt.Sprintf` sparingly**, as this can
+have a meaningful performance impact on a running production system. Consider
+and example where one may call `fmt.Sprintf` when logging something at **debug**
+level: even though an operator has configured their system to only log at
+**info** level and above, the expensive `fmt.Sprintf` calls will still take
+place, potentially slowing the system down. In such instances, consider printing
+values [lazily][log-lazy].
+
+### Examples of good log messages
+
+```golang
+// Operators generally wouldn't care whether an internal construct, like module
+// construction, has executed successfully.
+logger.Debug("Starting reactor", "module", "consensus")
+
+logger.Info("Committed block", "height", height, "appHash", appHash)
+
+// Include information about the error.
+logger.Error("Failed to execute important operation", "err", err)
+```
+
+### Examples of bad log messages
+
+```golang
+// Message starts with a lowercase letter, and will probably be called very
+// frequently, effectively spamming operators.
```

Comment:
Does this statement makes sense concerning operators ? 

On the `log levels` section it was mentioned that the `Debug` level should only be targeted to consensus engine developers, so should it still be a concern to operators even if the frequency is high ?

Reply:
Yeah I'll change this to `Info`.